File: backend/app/schemas/optimization.py
import uuid
from enum import StrEnum

from pydantic import BaseModel


class OptimizationStrategy(StrEnum):
    FREQUENCY = "frequency"
    HARMONIZED = "harmonized"
    BALANCED = "balanced"


class OptimizationRequest(BaseModel):
    sequence: str  # protein sequence
    organism_tax_id: int
    strategy: OptimizationStrategy = OptimizationStrategy.FREQUENCY
    avoid_restriction_sites: list[str] = []
    target_gc_min: float | None = None
    target_gc_max: float | None = None
    avoid_repeats: bool = True


class OptimizationResponse(BaseModel):
    job_id: uuid.UUID
    status: str
    optimized_sequence: str | None = None
    cai_before: float | None = None
    cai_after: float | None = None
    gc_content: float | None = None
    changes_summary: dict | None = None

File: frontend/src/types/optimization.ts
export enum OptimizationStrategy {
  FREQUENCY = "frequency",
  HARMONIZED = "harmonized",
  BALANCED = "balanced",
}

export interface OptimizationRequest {
  sequence: string;
  organism_tax_id: number;
  strategy: OptimizationStrategy;
  avoid_restriction_sites: string[];
  target_gc_min?: number;
  target_gc_max?: number;
  avoid_repeats: boolean;
}

export interface OptimizationResult {
  job_id: string;
  status: string;
  optimized_sequence?: string;
  cai_before?: number;
  cai_after?: number;
  gc_content?: number;
  changes_summary?: Record<string, unknown>;
}
